theano/gpuarray/multinomial.py

```diff
@@ -318,23 +325,29 @@ def gpu_kernels(self, node, name):
                     global_outs[n * outs_col_stride +
                                 c * outs_row_stride] = m;
 
-                    global_pvals_copy[m * pvals_col_stride + n * pvals_row_stride] = 0.0;
-                    cummul -= pvals_nm;
+                    if (! %(replace)s )
+                    {
+                        global_pvals_copy[m * pvals_col_stride + n * pvals_row_stride] = 0.0;
+                        cummul -= pvals_nm;
+                    }
                     done = true;
                 }
             }
             // No need to renormalize after the last samples.
             if (c == (n_samples - 1))
                 break;
-            // parallel renormalize the multinomial
-            for (ga_int k = LID_1; k < nb_outcomes; k+=LDIM_1)
+            if (! %(replace)s )
             {
-                global_pvals_copy[k * pvals_col_stride + n * pvals_row_stride] /= cummul;
+                // parallel renormalize the multinomial
+                for (ga_int k = LID_1; k < nb_outcomes; k+=LDIM_1)
```

Comment:
On line 473 change this line:

         if (nb_threads * nb_blocks < 2048)

to

         if (nb_threads * nb_blocks < 2048 && %(replace)d)

This will optimize the GPU usage as this part is skipped